jumble/integration/basic-flow.test.ts

```diff
@@ -15,106 +15,185 @@ import {
-    page = await browser!.newPage(FRONTEND_URL);
-    console.log(`Opened website at ${FRONTEND_URL}`);
-    await login(page);
-  });
-  afterEach(async () => {
-    await page!.close();
-  });
-  afterAll(async () => {
-    await browser!.close();
-  });
-
-  it("renders a new charm", async () => {
-    assert(page, "Page should be defined");
-    assert(testCharm, "Test charm should be defined");
-
-    await snapshot(page, "Initial state");
-
-    const anchor = await page.waitForSelector("nav a");
-    assert(
-      (await anchor.innerText()) === "common-knowledge",
-      "Logged in and Common Knowledge title renders",
-    );
-
-    await page.goto(
-      `${FRONTEND_URL}${testCharm.name}/${testCharm.charmId}`,
-    );
-    await snapshot(page, "Waiting for charm to render");
-
-    await waitForSelectorWithText(
-      page,
-      "a[aria-roledescription='charm-link']",
-      "Simple Value: 1",
-    );
-    await snapshot(page, "Charm rendered.");
-    assert(
-      true,
-      "Charm rendered successfully",
-    );
-
-    console.log("Clicking button");
-    // Sometimes clicking this button throws:
-    // https://jsr.io/@astral/astral/0.5.2/src/element_handle.ts#L192
-    // As if the reference was invalidated by a spurious re-render between
-    // getting an element handle, and clicking it.
-    await sleep(1000);
-    const button = await page.waitForSelector(
-      "div[aria-label='charm-content'] button",
-    );
-    await button.click();
-    await snapshot(page, "Button clicked");
-
-    console.log("Checking if title changed");
-    await waitForSelectorWithText(
-      page,
-      "a[aria-roledescription='charm-link']",
-      "Simple Value: 2",
-    );
-
-    await snapshot(page, "Title changed");
-
-    console.log("Inspecting charm to verify updates propagated from browser.");
-    const charm = await inspectCharm(
-      TOOLSHED_API_URL,
-      testCharm.name,
-      testCharm.charmId,
-    );
-
-    console.log("Charm:", charm);
-    assert(
-      charm.includes("Simple Value: 2"),
-      "Charm updates propagated.",
-    );
-  });
-
-  // Placeholder test ensuring browser can be used
-  // across multiple tests (replace when we have more integration tests!)
-  it("[placeholder]", () => {
-    assert(page, "Page should be defined");
-    assert(testCharm, "Test charm should be defined");
-  });
+let browser: Browser | void = undefined;
+let testCharm: { charmId: string; name: string } | void = undefined;
+
+Deno.test({
+  name: "integration tests",
+  fn: async (t) => {
+    let page: Page | void = undefined;
+    let failed = false;
+    const exceptions: string[] = [];
+
+    try {
+      failed = !await t.step({
+        name: "add charm via cli",
+        ignore: failed || exceptions.length > 0,
+        fn: async () => {
+          testCharm = await addCharm(TOOLSHED_API_URL);
+          console.log(`Charm added`, testCharm);
+        },
+      });
+
+      failed = !await t.step({
```

Comment:
Is this the idiomatic way to use `Deno.test`? I've never seen `!await` in my life!

Reply:
might not need to now - I was getting BDD hard killing after a minute .. I think
 
I'll try to switch back

Reply:
While making `deno-web-test`, discovered that e.g. `@std/testing/bdd` is implemented via `Deno.test((t) => {})`, where the nested BDD scopes are "steps" inside Deno.test

Reply:
That being said, I don't see the benefit of handrolling that